File: 1-500/0042. Trapping Rain Water.java
// Hard
// https://leetcode.cn/problems/trapping-rain-water/

import java.util.Stack;

class Solution1 {
    // Two Pointers -> more straightforward and visualize
    // Time:O(n), Space:O(1)
    public int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int leftMax = 0, rightMax = 0;
        int res = 0;

        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] >= leftMax) {
                    leftMax = height[left];
                } else {
                    res += leftMax - height[left];
                }
                left++;
            } else {
                if (height[right] >= rightMax) {
                    rightMax = height[right];
                } else {
                    res += rightMax - height[right];
                }
                right--;
            }
        }
        return res;
    }
}

class Solution2 {
    // Stack
    // Time:O(n), Space:O(n)
    public int trap(int[] height) {
        Stack<Integer> stack = new Stack<>();
        int res = 0;

        for (int i = 0; i < height.length; i++) {
            while (!stack.isEmpty() && height[i] > height[stack.peek()]) {
                int bottom = stack.pop();
                if (stack.isEmpty())
                    break;

                int left = stack.peek();
                int width = i - left - 1;
                int h = Math.min(height[left], height[i] - height[bottom]);
                res += width * h;
            }
            stack.push(i);
        }
        return res;
    }
}

File: 0001-0500/0042. Trapping Rain Water.py
# Hard
# Two Pointers
# Time:O(n), Space:O(1)
# https://leetcode.cn/problems/trapping-rain-water/

from typing import *

class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        
        left, right = 0, len(height) - 1
        left_max, right_max = 0, 0
        water = 0

        while left < right:
            if height[left] < height[right]:
                if height[left] >= left_max:
                    left_max = height[left]
                else:
                    water += left_max - height[left]
                left += 1 
            else:
                if height[right] >= right_max:
                    right_max = height[right]
                else:
                    water += right_max - height[right]
                right -= 1
        
        return water

# Can use stack too. Time:O(n), Space:O(n)
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        water = 0

        for i in range(len(height)):
            while stack and height[i] > height[stack[-1]]:
                top = stack.pop()

                if not stack:
                    break

                left = stack[-1]
                width = i - left - 1
                h = min(height[left], height[i]) - height[top]
                water += width * h
            stack.append(i)
        return water
    